File: sorting/quicksort.c
#include <stdio.h>

// void swap(int *i , int *j)
// {
//     int temp = *i;
//     *i = *j;
//     *j = temp;
// }

int partition(int arr[], int l, int h)
{
    int pivot = arr[l];
    int i = l;
    int j = h;

    while (i < j)
    {
        do
        {
            i++;
        } while (arr[i] <= pivot);
        do
        {
            j--;
        } while (arr[j] > pivot);
        if (i < j)
        {
            int temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }
    int temp = arr[l];
    arr[l] = arr[j];
    arr[j] = temp;
    return j;
}

void quicksort(int arr[], int l, int h)
{
    if (l < h)
    {
        int j = partition(arr, l, h);
        quicksort(arr, l, j);
        quicksort(arr, j + 1, h);
    }
}

int main(int argc, char const *argv[])
{
    int arr[] = {43, 13, 76, 23, 1, 12, 7, 6};
    int n = sizeof(arr)/sizeof(arr[0]);

    printf("Before sorting : ");
    for (int i = 0; i < n; i++)
    {
        printf("%d ", arr[i]);
    }

    quicksort(arr, 0, n-1);
    printf("\nAfter sorting : ");
    for (int i = 0; i < n; i++)
    {
        printf("%d ", arr[i]);
    }

    return 0;
}

File: sorting/quicksort.java
package sorting;

public class quicksort {

    static int count=0;
    static void swap(int a[], int i , int j )
    {
        int temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }

    static int partition(int a[], int p , int q)
    {
        int pivot = a[p];
        int i = p+1;
        int j = q;

        while (i<=j)
        {
            while (i<=j && a[i] <= pivot)
            {
                i++;
            }
            while (a[j] > pivot)
            {
                j--;
            }
            if (i<j)
            {
                swap(a, i, j);
                count ++;
            }
        }
        swap(a, p, j);
        count++;
        return j;

    }

    static void quick(int a[], int l , int h)
    {
        if (l<h)
        {
            int index = partition(a,l,h);
            quick(a,l, index);
            quick(a,index+1, h);

        }

    }
    static void print(int arr[],int n)
    {
        for (int i = 0; i < n; i++)
        {
            System.out.print(arr[i] + " ");
        }
    }

    public static void main(String[] args) {
        int arr[] = {54,21,5,1222,53,1323,765,43,34,10};
        int num = arr.length;
        System.out.println("Unsorted array : ");
        print(arr, num);
        System.out.println();
        quick(arr, 0, num-1);
        System.out.println("\nSorted array : ");
        print(arr, num);
        System.out.println("\nCount = "+count);

    }
}
